src/modules/logicModule/KPIRewarder.sol

```diff
@@ -0,0 +1,376 @@
+        override(StakingManager, OptimisticOracleIntegrator)
+        initializer
+    {
+        __Module_init(orchestrator_, metadata);
+
+        (
+            address stakingTokenAddr,
+            address currencyAddr,
+            address ooAddr,
+            uint64 liveness
+        ) = abi.decode(configData, (address, address, address, uint64));
+
+        __StakingManager_init(stakingTokenAddr);
+        __OptimisticOracleIntegrator_init(currencyAddr, ooAddr, liveness);
+    }
+
+    // ======================================================================
+    // View functions
+
+    /// @inheritdoc IKPIRewarder
+    function getKPI(uint KPInum) public view returns (KPI memory) {
+        return registryOfKPIs[KPInum];
+    }
+
+    /// @inheritdoc IKPIRewarder
+    function getAssertionConfig(bytes32 assertionId)
+        public
+        view
+        returns (RewardRoundConfiguration memory)
+    {
+        return assertionConfig[assertionId];
+    }
+
+    /// @inheritdoc IKPIRewarder
+    function getStakingQueue() public view returns (address[] memory) {
+        return stakingQueue;
+    }
+
+    // ========================================================================
+    // Assertion Manager functions:
+
+    /// @inheritdoc IKPIRewarder
+    /// @dev about the asserter address: any address can be set as asserter, it will be expected to pay for the bond on posting.
+    /// The bond tokens can also be deposited in the Module and used to pay for itself, but ONLY if the bond token is different from the one being used for staking.
+    /// If the asserter is set to 0, whomever calls postAssertion will be paying the bond.
+    function postAssertion(
+        bytes32 dataId,
+        bytes32 data,
+        address asserter,
+        uint assertedValue,
+        uint targetKPI
+    ) public onlyModuleRole(ASSERTER_ROLE) returns (bytes32 assertionId) {
+        // =====================================================================
+        // Input Validation
+
+        //  If the asserter is the Module itself, we need to ensure the token paid for bond is different than the one used for staking, since it could mess with the balances
+        if (
+            asserter == address(this)
+                && address(defaultCurrency) == stakingToken
+        ) {
+            revert Module__KPIRewarder__ModuleCannotUseStakingTokenAsBond();
+        }
+
+        // Make sure that we are targeting an existing KPI
+        if (KPICounter == 0 || targetKPI >= KPICounter) {
+            revert Module__KPIRewarder__InvalidKPINumber();
+        }
+
+        // Question: what kind of checks should or can we implement on the data side?
+        // Technically the value mentioned inside "data" (and posted publicly) wouldn't need to be the same as assertedValue...
+
+        // =====================================================================
+        // Staking Queue Management
+
+        for (uint i = 0; i < stakingQueue.length; i++) {
```

Comment:
We talked about the queue already. It is practically a POC, but at some point we might wanna do a rework if this becomes a problem

Reply:
Yeah, I agree. But for now it seems the most straightforward way to go with.